cmd/kubectl-gadget/filetop.go

```diff
@@ -0,0 +1,262 @@
+		var err error
+
+		nodeStats = make(map[string][]types.Stats)
+
+		if len(args) >= 1 {
+			interval, err = strconv.Atoi(args[0])
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "%q is not a valid value for interval\n", args[0])
+				os.Exit(1)
+			}
+		}
+
+		config := &utils.TraceConfig{
+			GadgetName:       "filetop",
+			Operation:        "start",
+			TraceOutputMode:  "Stream",
+			TraceOutputState: "Started",
+			CommonFlags:      &params,
+			Parameters: map[string]string{
+				"output_rows": strconv.Itoa(maxrows),
+				"interval":    strconv.Itoa(interval),
+				"sort_by":     sortby_str,
+			},
+		}
+
+		startprint()
+
+		if err := utils.RunTraceStreamCallback(config, callback); err != nil {
+			fmt.Fprintf(os.Stderr, "%s\n", err)
+			os.Exit(1)
+		}
+	},
+	SilenceUsage: true,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+		sortby, err = types.ParseSortBy(sortby_str)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	},
+	Args: cobra.MaximumNArgs(1),
+}
+
+func init() {
+	filetopCmd.Flags().IntVarP(&maxrows, "maxrows", "r", 20, "maximum rows to print")
+	filetopCmd.Flags().StringVarP(&sortby_str, "sort", "", "rbytes", "sort column")
+
+	rootCmd.AddCommand(filetopCmd)
+	utils.AddCommonFlags(filetopCmd, &params)
+}
+
+func callback(line string, node string) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	var event types.Event
+
+	if err := json.Unmarshal([]byte(line), &event); err != nil {
+		fmt.Fprintf(os.Stderr, "error unmarshalling json: %s", err)
+		return
+	}
+
+	if event.Error != "" {
+		fmt.Fprintf(os.Stderr, "error in node %s:%s", event.Node, event.Error)
+		return
+	}
+
+	nodeStats[node] = event.Stats
+}
+
+func startprint() {
+	ticker := time.NewTicker(time.Duration(interval) * time.Second)
+
+	go func() {
+		for {
+			print()
+			_ = <-ticker.C
+		}
+	}()
+}
+
+func print() {
+	// sort and print events
+	mu.Lock()
+
+	stats := []types.Stats{}
+	for _, stat := range nodeStats {
+		stats = append(stats, stat...)
+	}
+	nodeStats = make(map[string][]types.Stats)
+
+	mu.Unlock()
+
+	types.SortStats(stats, sortby)
+
+	switch params.OutputMode {
+	case utils.OutputModeColumns:
+		if term.IsTerminal(int(os.Stdout.Fd())) {
+			utils.ClearScreen()
+		} else {
+			fmt.Println("")
+		}
+		fmt.Printf("%-16s %-16s %-16s %-16s %-7s %-16s %-6s %-6s %-7s %-7s %1s %s\n",
+			"NODE", "NAMESPACE", "POD", "CONTAINER",
+			"TID", "COMM", "READS", "WRITES", "R_Kb", "W_Kb", "T", "FILE")
+		for idx, event := range stats {
+			if idx == maxrows {
+				break
+			}
+			fmt.Printf("%-16s %-16s %-16s %-16s %-7d %-16s %-6d %-6d %-7d %-7d %c %s\n",
+				event.Node, event.Namespace, event.Pod, event.Container,
+				event.Tid, event.Comm, event.Reads, event.Writes, event.ReadBytes/1024,
+				event.WriteBytes/1024, event.FileType, event.Filename)
+		}
+	case utils.OutputModeJson:
+		b, err := json.Marshal(stats)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error marshalling json: %v", err)
+			return
+		}
+		fmt.Println(string(b))
```

Comment:
I think it's good to have empty output there. Each interval some data is generated (it's empty if nothing happened. It can be useful if a tool is using that output to determine that not events happened (for cleaning the screen for instance).

Reply:
Ok, makes sense.